### PR TITLE
Add health check function

### PR DESCRIPTION
**Pull request by erimatnor**
A new health check function _timescaledb_internal.health() returns the
health and status of the database instance, including any configured
data nodes (in case the instance is an access node).

Since the function returns also the health of the data nodes, it tries
hard to avoid throwing errors. An error will fail the whole function
and therefore not return any node statuses, although some of the nodes
might be healthy.

The health check on the data nodes is a recursive (remote) call to the
same function on those nodes. Unfortunately, the check will fail with
an error if a connection cannot be established to a node (or an error
occurs on the connection), which means the whole function call will
fail. This will be addressed in a future change by returning the error
in the function result instead.

### CODE REVIEW

File: tsl/src/remote/healthcheck.c [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+			StringInfo cmd = makeStringInfo();
+			Oid fnamespaceid = get_func_namespace(fcinfo->flinfo->fn_oid);
+
+			appendStringInfo(cmd,
+							 "SELECT * FROM %s.%s()",
+							 get_namespace_name(fnamespaceid),
+							 get_func_name(fcinfo->flinfo->fn_oid));
+			data_node_list = data_node_get_node_name_list();
+			search_path = GetConfigOption("search_path", false, false);
+			result = ts_dist_cmd_invoke_on_data_nodes_using_search_path(cmd->data,
```
Comment (pmwkaa) [inline tsl/src/remote/healthcheck.c:111]: But this will fail and throw an error, right?

Comment (erimatnor) [reply tsl/src/remote/healthcheck.c:111]: Yes, exactly. This is pointed out in the commit message as future work.

**Comment by erimatnor** [comment]
This PR is missing tests, which needs to be added before merge.

File: tsl/src/remote/healthcheck.c [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+	_Anum_health_max,
+};
+
+#define Natts_health (_Anum_health_max - 1)
+
+static HeapTuple
+create_health_check_tuple(const char *data_node, bool in_recovery, TupleDesc tupdesc)
+{
+	Datum values[Natts_health];
+	bool nulls[Natts_health] = { true };
```
Comment (nikkhils) [inline tsl/src/remote/healthcheck.c:35]: better to use nulls[Natts_health] = {false};?

**Comment by erimatnor** [comment]
Tests added

**Review by nikkhils — APPROVED** [review]
LGTM!

File: tsl/src/remote/healthcheck.c [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+
+	if (NULL != data_node)
+	{
+		NameData *node_name = palloc(sizeof(NameData));
+		namestrcpy(node_name, data_node);
+		values[AttrNumberGetAttrOffset(Anum_health_node_name)] = NameGetDatum(node_name);
+		nulls[AttrNumberGetAttrOffset(Anum_health_node_name)] = false;
+	}
+
+	values[AttrNumberGetAttrOffset(Anum_health_healthy)] = BoolGetDatum(true);
```
Comment (pmwkaa) [inline tsl/src/remote/healthcheck.c:63]: So healthy column is reserved for the future use in such case? Like internal consistency check etc

File: tsl/src/remote/healthcheck.c [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+			StringInfo cmd = makeStringInfo();
+			Oid fnamespaceid = get_func_namespace(fcinfo->flinfo->fn_oid);
+
+			appendStringInfo(cmd,
+							 "SELECT * FROM %s.%s()",
+							 get_namespace_name(fnamespaceid),
+							 get_func_name(fcinfo->flinfo->fn_oid));
+			data_node_list = data_node_get_node_name_list();
+			search_path = GetConfigOption("search_path", false, false);
+			result = ts_dist_cmd_invoke_on_data_nodes_using_search_path(cmd->data,
```
Comment (pmwkaa) [inline tsl/src/remote/healthcheck.c:136]: not sure we need to set search_path here

File: tsl/src/remote/healthcheck.c [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+#include "dist_util.h"
+
+/*
+ * Functions and data structures for printing a health check result.
+ */
+enum Anum_show_conn
+{
+	Anum_health_node_name = 1,
+	Anum_health_healthy,
+	Anum_health_in_recovery,
```
Comment (pmwkaa) [inline tsl/src/remote/healthcheck.c:24]: Likely node being in_recovery should not be considered healthy as well...I guess it won't hurt to have extra details (ui maybe?), since we use it for AN as well

**Review by pmwkaa — APPROVED** [review]
looks good, just comments

File: tsl/src/remote/healthcheck.c (hunk 3 above)
Comment (erimatnor) [reply tsl/src/remote/healthcheck.c:63]: It would be `false` if, for example, the data node cannot establish a connection to the DN. But it is not possible to test now since it throws an error. It would be `true` if the node appears healthy, but, e.g., `in_recovery` is also `true`. We could, of course, simply use `false` in both cases, but I figured it might be useful information to also let the caller know that the node is recovering.

File: tsl/src/remote/healthcheck.c (hunk 4 above)
Comment (erimatnor) [reply tsl/src/remote/healthcheck.c:134]: Removed.

File: tsl/src/remote/healthcheck.c (hunk 3 above)
Comment (erimatnor) [reply tsl/src/remote/healthcheck.c]: Changed this so that healthy is `false` when `in_recovery` is true.

**Review by 65278 — APPROVED** [review]
LGTM. The PR isn't hard to follow, that said I have not run it.
